fix: DIA-1431: [FE] Storage type dropdown sometimes loads as blank (on Add/ Edit Source Storage modal)

nick-skriabin opened the pull request:
The list of storage types sometimes doesn't load. This PR attempts to address that by moving to a more robust way of loading storages.

This pull request includes changes to the `StorageSet` component in the `web/apps/labelstudio/src/pages/Settings/StorageSettings/StorageSet.jsx` file to improve the handling of storage data fetching and state management. The most important changes include replacing the `useState` and `useEffect` hooks with `useQuery` from `@tanstack/react-query`, using `useMemo` to compute loading and loaded states, and updating the rendering logic for storage cards.

Improvements to data fetching and state management:

* Replaced `useState` and `useEffect` hooks with `useQuery` from `@tanstack/react-query` for fetching `storageTypes` and `storages`.
* Updated the `fetchStorages` function to use `reloadStoragesList` and `reloadStorageTypes` from `useQuery` for re-fetching data.

Optimizations in rendering logic:

* Used `useMemo` to compute `loading` and `loaded` states based on `isLoading` and `isSuccess` from `useQuery`.
* Updated the rendering logic to check `storagesLoaded` before mapping over `storages` to render `StorageCard` components.

<!--

This description MUST be filled out for a PR to receive a review. Its primary purposes are:

 - to enable your reviewer to review your code easily, and
 - to convince your reviewer that your code works as intended.

Some pointers to think about when filling out your PR description:
 - Reason for change: Description of problem and solution
 - Screenshots: All visible changes should include screenshots.
 - Rollout strategy: How will this code be rolled out? Feature flags / env var / other
 - Testing: Description of how this is being verified
 - Risks: Are there any known risks associated with this change, eg to security or performance?
 - Reviewer notes: Any info to help reviewers approve the PR
 - General notes: Any info to help onlookers understand the code, or callouts to significant portions.

You may use AI tools such as Copilot Actions to assist with writing your PR description (see https://docs.github.com/en/copilot/using-github-copilot/using-github-copilot-for-pull-requests/creating-a-pull-request-summary-with-github-copilot); however, an AI summary isn't enough by itself. You'll need to provide your reviewer with strong evidence that your code works as intended, which requires actually running the code and showing that it works.

-->

nick-skriabin committed:
Migrate to react query

nass600 commented on web/apps/labelstudio/src/pages/Settings/StorageSettings/StorageSet.jsx:
```diff
@@ -1,60 +1,73 @@
+  const storageTypesQueryKey = ["storage-types", target];
+  const storagesQueryKey = ["storages", target, project?.id];
+
+  const {
```
I would separate concerns here and put the API layer via React Query as hooks in the hooks folder

nick-skriabin committed:
Separate concerns

nass600 commented on web/apps/labelstudio/src/pages/Settings/StorageSettings/hooks/useStorageCard.tsx:
```diff
@@ -0,0 +1,74 @@
+import { useQuery } from "@tanstack/react-query";
```
I'll try to go more modular:

```
import { useQuery } from '@tanstack/react-query'
import { useAPI } from 'apps/labelstudio/src/providers/ApiProvider'
import { useCallback } from 'react'

export function useStorageTypes(target: 'import' | 'export') {
  const api = useAPI()
  return useQuery({
    queryKey: ['storage-types', target],
    queryFn: async () => {
      const result = await api.callApi('storageTypes', {
        params: { target },
        errorFilter: () => true,
      })
      return result?.$meta.ok ? result : []
    },
  })
}

export function useStorages(target: 'import' | 'export', projectId?: number, storageTypesLoaded?: boolean) {
  const api = useAPI()
  return useQuery({
    queryKey: ['storages', target, projectId],
    queryFn: async () => {
      if (!storageTypesLoaded) return []
      const result = await api.callApi('listStorages', {
        params: { project: projectId, target },
        errorFilter: () => true,
      })
      return result?.$meta.ok ? result : []
    },
    enabled: storageTypesLoaded,
  })
}

export function useStorageCard(target: 'import' | 'export', projectId?: number) {
  const storageTypesQuery = useStorageTypes(target)
  const storagesQuery = useStorages(target, projectId, storageTypesQuery.isSuccess)

  const fetchStorages = useCallback(async () => {
    await storageTypesQuery.refetch()
    await storagesQuery.refetch()
  }, [storageTypesQuery.refetch, storagesQuery.refetch])

  return {
    ...storageTypesQuery,
    ...storagesQuery,
    fetchStorages,
  }
}

```

nick-skriabin committed:
Move loading state to hook

nick-skriabin committed:
Modular hook

nick-skriabin committed:
Add disabled state to the button when loading storages

AndrejOros commented:
/fm sync 
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14238056249)

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-dia-1431/cloud-storage-types'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14238056249

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14238056249
